### PR TITLE
need to call as.POSIXct() with tz after "TZ" ENV VAR gets changed

### PR DESCRIPTION
**Pull request by shrektan**
Otherwise, it will fail on macOS. See https://github.com/Rdatatable/data.table/pull/4261#issuecomment-592484510 for the details.

---

Update: I've reported this on https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17724

---

Update: As Luke Tierney points out (in the above URL), unset the env var by calling `Sys.setenv(TZ="")` is wrong. The correct way is to call `Sys.unsetenv("TZ")`.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
@jangorecki can we add `TZ=Asia/Jakarta` (or any non-UTC tz) to one of the GL CI test suites?

Maybe also something `LANGUAGE=es_AR` to check sensitivity to locale as well

**Comment by jangorecki** [comment]
@MichaelChirico windows builds runs on
```
TZ=America/Los_Angeles, locale='LC_COLLATE=C;LC_CTYPE=English_United States.1252;LC_MONETARY=English_United States.1252;LC_NUMERIC=C;LC_TIME=English_United States.1252'
```
https://gitlab.com/Rdatatable/data.table/-/jobs/443985937/artifacts/file/bus/test-rel-win/data.table.Rcheck/tests_x64/main.Rout

**Comment by MichaelChirico** [comment]
@jangorecki I see, good to know. Is that a property of the docker image? I don't see anything in `.gitlab-ci.yml` about it

In any case, it appears we missed this issue anyway -- would it make sense to add a different TZ & locale to one of the other builds?

@shrektan what's your `Sys.timezone()` / locale info?

**Comment by shrektan** [comment]
### Sys.timezone()

```
[1] "Asia/Shanghai"
```

### error related

```
Error in test.data.table() : 
  6 errors out of 9867. Search tests/tests.Rraw for test numbers: 2137.04, 2137.06, 2137.07, 2137.08, 2137.09, 2137.1.
```

### Code that below comment uses

```r
library(data.table)
DT = data.table(hour31   = as.ITime(seq(as.POSIXct("2020-01-01 07:00:40"), by = "31 min", length.out = 9)),
                hour30   = as.ITime(seq(as.POSIXct("2020-01-01 07:00:00"), by = "30 min", length.out = 9)),
                minute31 = as.ITime(seq(as.POSIXct("2020-01-01 07:00:00"), by = "31 sec", length.out = 9)),
                minute30 = as.ITime(seq(as.POSIXct("2020-01-01 07:00:00"), by = "30 sec", length.out = 9)))
DT
```

### An interesting observation is:

Before I run `test.data.table()`, the above code returns:

```
     hour31   hour30 minute31 minute30
1: 07:00:40 07:00:00 07:00:00 07:00:00
2: 07:31:40 07:30:00 07:00:31 07:00:30
3: 08:02:40 08:00:00 07:01:02 07:01:00
4: 08:33:40 08:30:00 07:01:33 07:01:30
5: 09:04:40 09:00:00 07:02:04 07:02:00
6: 09:35:40 09:30:00 07:02:35 07:02:30
7: 10:06:40 10:00:00 07:03:06 07:03:00
8: 10:37:40 10:30:00 07:03:37 07:03:30
9: 11:08:40 11:00:00 07:04:08 07:04:00
```
After `test.data.table()`, the result becomes:

```
     hour31   hour30 minute31 minute30
    <ITime>  <ITime>  <ITime>  <ITime>
1: 15:00:40 15:00:00 15:00:00 15:00:00
2: 15:31:40 15:30:00 15:00:31 15:00:30
3: 16:02:40 16:00:00 15:01:02 15:01:00
4: 16:33:40 16:30:00 15:01:33 15:01:30
5: 17:04:40 17:00:00 15:02:04 15:02:00
6: 17:35:40 17:30:00 15:02:35 15:02:30
7: 18:06:40 18:00:00 15:03:06 15:03:00
8: 18:37:40 18:30:00 15:03:37 15:03:30
9: 19:08:40 19:00:00 15:04:08 15:04:00
```

### What's more interesting is:

If I add a `browser()` into the `test.Rraw`  file, the execution will stop in the middle, and if I choose to continue, the error will not occur. However, if I remove the `browser()` line, the error will occur.

Again, I checked both `Sys.getlocale()` and `Sys.timezone()`, they returned the same, as if never changed.

**Comment by shrektan** [comment]
See this weird issue...

### Before it returns expected

```r
d> as.POSIXct("2020-01-01 07:00:40") 
[1] "2020-01-01 07:00:40 CST"
```

### After I run `test.data.table()` error throws and the same code returns differently


image

**Comment by jangorecki** [comment]
@MichaelChirico no, windows builds runs not in a docker but using ssh executor on windows OS directly.

**Comment by shrektan** [comment]
OK I just find the root causes:

https://github.com/Rdatatable/data.table/blob/b1b1832b0d2d4032b46477d9fe6efb15006664f4/inst/tests/tests.Rraw#L16554-L16560

I'm not sure it's a bug or not but we just need to call `as.POSIXct` with `tz` explicitly after we change the `TZ` env variable.

``` r
# system timezone is not usually UTC, so as.ITime.POSIXct shouldn't assume so, #4085
as.POSIXct("2020-01-01 07:00:40")
#> [1] "2020-01-01 07:00:40 CST"
oldtz=Sys.getenv('TZ')
Sys.setenv(TZ='Asia/Jakarta') # UTC+7
as.POSIXct("2020-01-01 07:00:40")
#> [1] "2020-01-01 07:00:40 WIB"
Sys.setenv(TZ=oldtz)
as.POSIXct("2020-01-01 07:00:40")
#> [1] "2020-01-01 15:00:40 CST"
as.POSIXct("2020-01-01 07:00:40", tz = Sys.timezone())
#> [1] "2020-01-01 07:00:40 CST"
as.POSIXct("2020-01-01 07:00:40")
#> [1] "2020-01-01 07:00:40 CST"
```

<sup>Created on 2020-02-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by MichaelChirico** [comment]
on `rocker/r-base` I see:

```
Sys.timezone()
# [1] "Etc/UTC"
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 UTC"
oldtz=Sys.getenv('TZ')
Sys.setenv(TZ='Asia/Jakarta') # UTC+7
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 WIB"
Sys.setenv(TZ=oldtz)
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 UTC"
as.POSIXct("2020-01-01 07:00:40", tz = Sys.timezone())
# [1] "2020-01-01 07:00:40 UTC"
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 UTC"
```

On my Mac I see:

```
Sys.timezone()
# [1] "Asia/Kuala_Lumpur"
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 15:00:40 +08"
oldtz=Sys.getenv('TZ')
Sys.setenv(TZ='Asia/Jakarta') # UTC+7
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 WIB"
Sys.setenv(TZ=oldtz)
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 15:00:40 +08"
as.POSIXct("2020-01-01 07:00:40", tz = Sys.timezone())
# [1] "2020-01-01 07:00:40 +08"
as.POSIXct("2020-01-01 07:00:40")
# [1] "2020-01-01 07:00:40 +08"
```

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -16552,14 +16552,15 @@ dt = data.table(SomeNumberA=c(1,1,1),SomeNumberB=c(1,1,1))
 test(2123, dt[, .(.N, TotalA=sum(SomeNumberA), TotalB=sum(SomeNumberB)), by=SomeNumberA], data.table(SomeNumberA=1, N=3L, TotalA=1, TotalB=3))
 
 # system timezone is not usually UTC, so as.ITime.POSIXct shouldn't assume so, #4085
-oldtz=Sys.getenv('TZ')
+oldtz = Sys.getenv('TZ', unset = NA)
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw]: nice catch, h/t to Luke

**Comment by mattdowle** [comment]
Linking to #4088 and #4085 might may have been involved.
It seems that the changes in this PR are already in `master`. Trying to link to when that happened, and then will close ...

**Pull request merged**